### PR TITLE
Add basic E2E WAF tests

### PR DESCRIPTION
**Pull request by ArSn**
Add some basic WAF tests to check if requests are blocked when they should.

#### Changes proposed in this Pull Request:

* Add the spec for WAF e2e test
* Add simple test for request blocking

#### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion

While detailed tests of the WAF engine are better suited to unit tests, and also need access to the full set of rules which are not present here, we still believe a single E2E test to ensure that requests are correctly blocked when the module is enabled is valuable.

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* See CI checks
* Run locally with `$ pnpm test:run specs/waf/blocking.test.js --headed` from within `projects/plugins/jetpack/tests/e2e`

### CODE REVIEW

File: .github/files/create-e2e-projects-matrix.sh
Comment (brbrr) [inline .github/files/create-e2e-projects-matrix.sh]: Do we need to add a separate CI action here?

File: tools/e2e-commons/plugins/e2e-plan-data-interceptor.php
Comment (brbrr) [inline tools/e2e-commons/plugins/e2e-plan-data-interceptor.php]: It doesn't look like an easily maintainable approach. Is there anything else we can do here? 

Also, it looks like super specific for the WAF test, so maybe we can isolate it somehow?

File: projects/plugins/jetpack/tests/e2e/specs/waf/blocking.test.js
Comment (brbrr) [inline projects/plugins/jetpack/tests/e2e/specs/waf/blocking.test.js]: Is there a particular reason to use `parallel` here?

File: projects/plugins/jetpack/tests/e2e/specs/waf/blocking.test.js
Comment (brbrr) [inline projects/plugins/jetpack/tests/e2e/specs/waf/blocking.test.js]: I'm having a hard time understanding the meaning behind this assertion. Can you please elaborate?

File: tools/e2e-commons/pages/page-actions.js
Comment (brbrr) [inline tools/e2e-commons/pages/page-actions.js:32]: Is this intentional?

File: .github/files/create-e2e-projects-matrix.sh
Comment (ArSn) [reply .github/files/create-e2e-projects-matrix.sh]: Well, there weren't any WAF e2e tests before, I guess we could include them in some of the other directories, but they all are pretty much different topics. Where else would we best group it under?

File: tools/e2e-commons/plugins/e2e-plan-data-interceptor.php
Comment (ArSn) [reply tools/e2e-commons/plugins/e2e-plan-data-interceptor.php]: The intent of the test is to check that it "still blocks at all", so we only really need that one rule. Also this is our workaround for not creating an actual subscription.

We thought about moving that to a different file and register it on the same filter, but felt like it was technically also "data to do with the plan", so we left it in here. Would you think moving it to a different file would improve it?

File: projects/plugins/jetpack/tests/e2e/specs/waf/blocking.test.js
Comment (ArSn) [reply projects/plugins/jetpack/tests/e2e/specs/waf/blocking.test.js]: No, we probably oriented that too exactly on other examples in the repo :)

File: projects/plugins/jetpack/tests/e2e/specs/waf/blocking.test.js
Comment (ArSn) [reply projects/plugins/jetpack/tests/e2e/specs/waf/blocking.test.js]: The job of the WAF is to block certain requests, and that is what we are testing here. Given that when a request is blocked, the code does `die()` with a specific message, we never render the page. The assertion is just to ensure that we indeed to not see a page rendered in the output.

File: tools/e2e-commons/pages/page-actions.js
Comment (ArSn) [reply tools/e2e-commons/pages/page-actions.js:32]: Yes.

We are relying on reading data from the response (the http status and the body), and the response gets returned from `this.page.goto`, but was previously just tossed here.

We hoped not to break any other tests with it and according to the checks on this PR we didn't.

Do you think it will be an issue?

File: projects/plugins/jetpack/tests/e2e/specs/waf/blocking.test.js
Comment (ArSn) [reply projects/plugins/jetpack/tests/e2e/specs/waf/blocking.test.js]: Just to be sure before I change it: Is it problematic that we call this parallel? I'm thinking that someone adding tests to this later on may forget to do so when it would indeed be a good idea to save execution time?

File: tools/e2e-commons/pages/page-actions.js
Comment (brbrr) [reply tools/e2e-commons/pages/page-actions.js:32]: Oh, right. I didn't realize that `page.goto` actually returns a meaningful/useful result. All good then :)

File: projects/plugins/jetpack/tests/e2e/specs/waf/blocking.test.js
Comment (brbrr) [reply projects/plugins/jetpack/tests/e2e/specs/waf/blocking.test.js]: Ok, gotcha. If there is no other way to assert that, maybe we can have a comment explaining the behavior here? It would be useful for folks looking into that test without a good understanding of the WAF implementation.

File: projects/plugins/jetpack/tests/e2e/specs/waf/blocking.test.js
Comment (brbrr) [reply projects/plugins/jetpack/tests/e2e/specs/waf/blocking.test.js]: > Is it problematic that we call this parallel?

Not at all. At least at this point. When running tests in parallel, the same site state (connection, plan, browser session etc) will be reused. If one of the tests will modify that (i.e. you'll add a WAF test for not logged in user), it will affect other tests in the same spec file.

File: tools/e2e-commons/plugins/e2e-plan-data-interceptor.php
Comment (brbrr) [reply tools/e2e-commons/plugins/e2e-plan-data-interceptor.php]: The "problem" here is that this plugin is part of a shared E2E infra and was created to help with plan mocking. Your change is quite specific to Jetpack plugin. It wouldn't break tests for other plugins, but it just messes up the separation of context. Ideally, this added logic should live in Jetpack E2Es folder, but since we don't have the logic to load functional plugins from within project folders, it should be good for now. @adimoldovan, what do you think?

File: .github/files/create-e2e-projects-matrix.sh
Comment (brbrr) [reply .github/files/create-e2e-projects-matrix.sh]: Our CI matrix is organized with efficiency in mind. `pos-connection` probably would be the best place for them.

The problem with additional action is that it will increase the total number of actions being run for a PR, which is something I'd like to avoid if possible. Also, the test itself should be fast enough, not to slow down the `post-connection` action significantly.

File: tools/e2e-commons/plugins/e2e-plan-data-interceptor.php
Comment (adimoldovan) [reply tools/e2e-commons/plugins/e2e-plan-data-interceptor.php]: I would move this into a different file to be honest. 
It will also make things easier for when we'll have the logic to load functional plugins from within project folders (hopefully soon enough)

**Comment by brbrr** [comment]
So I was thinking about the justification of the need to have an E2E test for WAF functionality. As it was discussed internally, I don't think it makes sense to have a full-blown E2E test when we are really testing a backend library here. Am I missing something?

**Comment by ArSn** [comment]
> So I was thinking about the justification of the need to have an E2E test for WAF functionality. As it was discussed internally, I don't think it makes sense to have a full-blown E2E test when we are really testing a backend library here. Am I missing something?

We are currently in the process of adding a bunch of tests that will indeed be run at the library level, specifically to ensure that certain specific requests still go through (are not blocked). We'll also add tests for things that will be blocked on the library level, though.

The reasoning behind having that one very simple e2e test was that, if we did ALL the testing on the library level and limit ourselves to that, we might be running into a situation where we think everything is fine, but in actuality the WAF never even blocks anything. This is where the e2e test comes into play, just to verify that it does indeed block a real request in a close-to-real-world-as-possible situation.

**Comment by ArSn** [comment]
P.S.: Don't get me wrong, I think we all agree that the e2e test is indeed cumbersome to have and to maintain, possibly too. It would be the easy way out to just not have it, but to deliver great quality, we just want to ensure that our product does as it should 🙂

File: tools/e2e-commons/plugins/e2e-plan-data-interceptor.php
Comment (ArSn) [reply tools/e2e-commons/plugins/e2e-plan-data-interceptor.php]: Moved it to a separate file 🙂 Thanks for your input!

File: .github/files/create-e2e-projects-matrix.sh
Comment (ArSn) [reply .github/files/create-e2e-projects-matrix.sh]: From the looks of it, the tests in "post-connection" provision a Free plan. Does that make any difference to the reasoning behind putting it in the same group? Or should we put it in there regardless so that we still save the extra CI job?

**Comment by brbrr** [comment]
> we might be running into a situation where we think everything is fine, but in actually the WAF never even blocks anything

Yeah, it makes sense. From what I understand, you basically want to test `die` call in the WAF library. I guess it depends on how testable the implementation is, but I think it should be possible to cover this use case with unit tests. 

Also, I see that you are mocking the plan and WAF endpoint data, which means this is already kinda an integration test. Would it make sense to actually rely on real data in this case?

**Comment by ArSn** [comment]
> Yeah, it makes sense. From what I understand, you basically want to test `die` call in the WAF library.

Yes, for that, we'll also have unit tests. But the point is that we still want to make sure that this actually does its job when used in the jetpack plugin and with a request that is to be blocked.

> Also, I see that you are mocking the plan and WAF endpoint data, which means this is already kinda an integration test. Would it make sense to actually rely on real data in this case?

Absolutely! And I think we would very much like to do so in the future. It just didn't work like that out of the box, and it seems there is more work to do until the tests can actually do that. Admittedly, we are at a meetup right and tried to make it small enough to be able to ship before we go home 🙂

File: .github/files/create-e2e-projects-matrix.sh
Comment (brbrr) [reply .github/files/create-e2e-projects-matrix.sh]: > From the looks of it, the tests in "post-connection" provision a Free plan.

That's true, but it should be ok since these tests are actually mocking plans and running sequentially.

Comment (ArSn) [reply .github/files/create-e2e-projects-matrix.sh]: Moved it into the "post-connection" group :)

**Review by brbrr — DISMISSED** [review]
Looks good to me! Before we merge this, let's get a review from your team as well.

**Comment by brbrr** [comment]
Hi @Automattic/yamato-scan. Can you take a look at this PR?

**Review by jwidavid — DISMISSED** [review]
Everything looks good - I was not able to successfully run the test in my environment, but I was present when we got the test working at the meetup.

Perhaps @ArSn could walk me through the process of getting that working when he gets back from AFK. Also, just for posterity, there is a very minor documentation typo on line 34 of `waf-blocking.test.js`

File: projects/plugins/jetpack/tests/e2e/specs/post-connection/waf-blocking.test.js
Comment (brbrr) [inline projects/plugins/jetpack/tests/e2e/specs/post-connection/waf-blocking.test.js]:
```suggestion
			The assertion is just to ensure that we indeed do not see a page rendered in the output.
```

h/t to @jwidavid

**Review by brbrr — APPROVED** [review]

**Pull request merged**
